File: packages/ogdf.js/src/layouts/energybased/segv.js
import { PARAMETER_TYPE } from '../../utils/parameters'
import createLayout from '../creater'

const NAME = 'SEGV'
const OUR_PARAMETERS = {
    useWorker: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false,
    },
}

const ORIGIN_PARAMETERS = {
    avgConvergenceFactor: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.1,
    },
    forceModel: {
        type: PARAMETER_TYPE.CATEGORICAL,
        range: ["FruchtermanReingold", "FruchtermanReingoldModAttr", "FruchtermanReingoldModRep", "Eades", "Hachul", "Gronemann"],
        default: "FruchtermanReingold",
    },
    forceModelImprove: {
        type: PARAMETER_TYPE.CATEGORICAL,
        range: ["FruchtermanReingold", "FruchtermanReingoldModAttr", "FruchtermanReingoldModRep", "Eades", "Hachul", "Gronemann"],
        default: "FruchtermanReingoldModRep",
    },
    idealEdgeLength: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 20.0 + Math.hypot(20.0, 20.0),
    },
    iterations: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 400,
    },
    iterationsImprove: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 200,
    },
    maxConvergenceFactor: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.2,
    },
    maxThreads: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 3,
    },
    minDistCC: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 30.0,
    },
    noise: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: true,
    },
    pageRatio: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 1.0,
    },
    scaleFunctionFactor: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 4.0,
    },
    scaling: {
        type: PARAMETER_TYPE.CATEGORICAL,
        range: ["input", "userBoundingBox", "scaleFunction", "useIdealEdgeLength"],
        default: "scaleFunction",
    },
    userBoundingBoxXmin: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.0,
    },
    userBoundingBoxYmin: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.0,
    },
    userBoundingBoxXmax: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 100.0,
    },
    userBoundingBoxYmax: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 100.0,
    },
}

const segv = createLayout(NAME, OUR_PARAMETERS, ORIGIN_PARAMETERS, {})
export default segv

File: packages/ogdf.js/src/entry/energybased/segv.cpp
#include <ogdf/energybased/SpringEmbedderGridVariant.h>
#include "../main.h"

EM_PORT_API(float *) SEGV(int node_num, int link_num, int* source, int* target, double avgConvergenceFactor, int forceModel, int forceModelImprove, double idealEdgeLength, int iterations, int iterationsImprove, double maxConvergenceFactor, unsigned int maxThreads, double minDistCC, bool noise, double pageRatio, double scaleFunctionFactor, int scaling, double userBoundingBoxXmin, double userBoundingBoxYmin, double userBoundingBoxXmax, double userBoundingBoxYmax) {
	node* nodes;
	Graph G;
	GraphAttributes GA(G, GraphAttributes::nodeGraphics | GraphAttributes::edgeGraphics);

	nodes = new node[node_num];
	for (int i = 0; i < node_num; i++){
		nodes[i] = G.newNode();
	}

	edge e;

	for (int i = 0; i < link_num; i++) {
		e = G.newEdge(nodes[source[i]], nodes[target[i]]);
		GA.bends(e);
	}
	//LayoutModule
	SpringEmbedderGridVariant *model = new SpringEmbedderGridVariant();
	
	model->avgConvergenceFactor(avgConvergenceFactor);
    model->forceModel(static_cast<SpringForceModel>(forceModel));
    model->forceModelImprove(static_cast<SpringForceModel>(forceModelImprove));
    model->idealEdgeLength(idealEdgeLength);
    model->iterations(iterations);
    model->iterationsImprove(iterationsImprove);
    model->maxConvergenceFactor(maxConvergenceFactor);
    model->maxThreads(maxThreads);
    model->minDistCC(minDistCC);
    model->noise(noise);
    model->pageRatio(pageRatio);
    model->scaleFunctionFactor(scaleFunctionFactor);
    model->scaling(static_cast<spring_embedder::SpringEmbedderBase::Scaling>(scaling));
    model->userBoundingBox(userBoundingBoxXmin,userBoundingBoxYmin,userBoundingBoxXmax,userBoundingBoxYmax);

    model->call(GA);

	float* re = (float*)malloc(node_num * 2 * 4);
    for(int i = 0; i < node_num; ++i) {
        re[i * 2] = GA.x(nodes[i]);
        re[i * 2 + 1] = GA.y(nodes[i]);
    }

    return re;
}